### PR TITLE
Add insertMany interface for IColumn

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #8904

Problem Summary:

### What is changed and how it works?

I used `insertManyFrom` in https://github.com/pingcap/tiflash/pull/8854, where the source column is a const column while the target column is not and it causes #8904. To fix this, add `insertMany` interface to avoid convert const column to full column.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Columns/IColumn.h [hunk 1]
```diff
@@ -130,18 +130,29 @@ class IColumn : public COWPtr<IColumn>
...
     /// Appends one element from other column with the same type multiple times.
+    /// Note: the source column and the destination column must be of the same type, can not ColumnXXX->insertManyFrom(ConstColumnXXX, ...)
     virtual void insertManyFrom(const IColumn & src, size_t position, size_t length) = 0;
 
     /// Appends disjunctive elements from other column with the same type.
+    /// Note: the source column and the destination column must be of the same type, can not ColumnXXX->insertDisjunctFrom(ConstColumnXXX, ...)
     virtual void insertDisjunctFrom(const IColumn & src, const std::vector<size_t> & position_vec) = 0;
 
+    /// Appends one field multiple times. Can be optimized in inherited classes.
+    virtual void insertMany(const Field & field, size_t length)
```
Comment (yibin87) [inline dbms/src/Columns/IColumn.h:150]: May reduce virtual function call like this:
https://github.com/pingcap/tiflash/blob/adc57e22149d04d76d26f5193134a05a938af7c7/dbms/src/Columns/IColumn.h#L476-L477

**Review by yibin87 — APPROVED** [review]
Others LGTM

File: dbms/src/Columns/IColumn.h (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Columns/IColumn.h:150]: It should be called as `IColumn->insertMany(v)`

Comment (yibin87) [reply dbms/src/Columns/IColumn.h:150]: Yeah, the insert virutal function is called inside the loop

Comment (Lloyd-Pottiger) [reply dbms/src/Columns/IColumn.h:150]: compiler can help optimize this, no vtable lookup here https://gcc.godbolt.org/z/Y48rMoqqc

File: dbms/src/Columns/ColumnVector.h [hunk 2]
```diff
@@ -229,6 +229,11 @@ class ColumnVector final : public COWPtrHelper<ColumnVectorHelper, ColumnVector<
             data[i + old_size] = src_container[position_vec[i]];
     }
 
+    void insertMany(const Field & field, size_t length) override
+    {
+        data.resize_fill(data.size() + length, static_cast<T>(field.get<T>()));
+    }
+
```
Comment (JaySon-Huang) [inline dbms/src/Columns/ColumnVector.h:236]: /cc @breezewish vector may need to pick it back to the cse release-7.5 branch

File: dbms/src/Columns/IColumn.h (hunk 1 above)
Comment (yibin87) [reply dbms/src/Columns/IColumn.h:150]: I don't think so, think like this: different derived instances can call insertMany function, thus optimizer can't infer which insert function to be called next. Thus optimizer should lookup the vtable of current instance to find the actual insert function address. That's what "call    qword ptr [rax + 8]" do, I guess.

Comment (yibin87) [reply dbms/src/Columns/IColumn.h:150]: Anyway, I think it is ok to improve it later.

File: dbms/src/Columns/tests/gtest_column_insertFrom.cpp [hunk 3]
```diff
@@ -57,54 +57,82 @@ class TestColumnInsertFrom : public ::testing::Test
...
-            ColumnWithTypeAndName result(std::move(cols[1]), col_with_type_and_name.type, "");
-            compareColumn(ref, result);
+            for (size_t i = 0; i < 2; ++i)
+                cols[i] = column_ptr->cloneEmpty();
+            for (size_t i = 0; i < 6; ++i)
+                cols[0]->insertFrom(*column_ptr, 1);
+            if unlikely (
+                typeid_cast<const ColumnNothing *>(column_ptr.get())
+                || typeid_cast<const ColumnSet *>(column_ptr.get()))
+            {
```
Comment (JaySon-Huang) [inline dbms/src/Columns/tests/gtest_column_insertFrom.cpp:125]:
```suggestion
            if (unlikely (
                typeid_cast<const ColumnNothing *>(column_ptr.get())
                || typeid_cast<const ColumnSet *>(column_ptr.get())))
            {
```

Better use `if ( unlikely(...) )`. `( ... )` is needed

File: dbms/src/Columns/ColumnVector.h (hunk 2 above)
Comment (JaySon-Huang) [reply dbms/src/Columns/ColumnVector.h:236]: `insertMany` is a new interface. I misunderstand it. We don't need to backport this in the cse branch.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Columns/ColumnVector.h (hunk 2 above)
Comment (breezewish) [reply dbms/src/Columns/ColumnVector.h:236]: Just curious,

> I used insertManyFrom in https://github.com/pingcap/tiflash/pull/8854, where the source column is a const column while the target column is not and it causes https://github.com/pingcap/tiflash/issues/8904.

Why insertManyFrom causes issues?

Comment (JaySon-Huang) [reply dbms/src/Columns/ColumnVector.h:236]: Actually, it is caused by calling `ColumnVector::insertManyForm(ColumnConst, ... )`. So `static_cast<const Self &>(column_const)` just returns a random memory address and introduces bugs.

https://github.com/pingcap/tiflash/blob/8cec573af26adf5d3d974506ac4e9cbaecd7d5f6/dbms/src/Columns/ColumnVector.h#L216-L220

**Comment by yibin87** [comment]
Caused by my tidb PR, https://github.com/pingcap/tidb/pull/52143, I'm fixing it now. Maybe you can temporarily bypass this case to pass the integration tests.

**Pull request merged**